File: jsrc/machine/SxxMachine/PredTable.java
package SxxMachine;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Map.Entry;

public class PredTable {

	public static Map<String, Operation> predicateCache = new HashMap<String, Operation>();
	public static Map<String, Operation> hiddenCache = new HashMap<String, Operation>();
	public static Map<String, Operation> initsCache = new HashMap<String, Operation>();

	public static void registerPredicate(String pkg, String functor, int arity, Operation cont) {
		registerPredicate(pkg, functor, arity, cont, false);
	}

	public static void registerPredicate(String pkg, String functor, int arity, Operation cont, boolean forced) {
		if (!storePred(keyString(pkg, functor, arity), cont, forced)) {
		}
		if (!storePred(keyString(null, functor, arity), cont, false)) {
		}
	}

	public static boolean storePred(String key, Operation cont, boolean forced) {
		Operation op = predicateCache.get(key);
		if (op == null) {
			predicateCache.put(key, cont);
			return true;
		} else if (op == cont) {
			return true;
		}
		String oldmodule = moduleFromClass(op);
		if (cont == null) {

			if (forced) {
				System.err.println("Deleting " + key + " which is implemented at " + oldmodule);
				predicateCache.remove(key);
				return true;
			} else {
				System.err.println("Not deleting " + key + " which is implemented at " + oldmodule);
				return false;
			}
		}
		String newmodule = moduleFromClass(cont);
		{
			if (op.getClass() != op.getClass()) {
				if (!forced) {
					System.err.println(
							"Not overriding " + key + " which is implemented at " + oldmodule + " with " + newmodule);
					hiddenCache.put(key, cont);
					return false;
				} else {
				}
			}
			hiddenCache.put(key, op);
			predicateCache.put(key, cont);
			System.err.println("Overriding " + key + " which is implemented at " + oldmodule + " with " + newmodule);
			return true;
		}
	}

	public static void registerWMangle(String pkg, String functor, String mangle, int arity, Operation cont) {
		registerPredicate(pkg, functor, arity, cont);
		registerPredicate(pkg, mangle, arity, cont);
	}

	private static String keyString(String pkg, String functor, int arity) {
		if (pkg == null || isGlobal(pkg))
			return PredFA(functor, arity);
		return pkg + "::" + PredFA(functor, arity);
	}

	/**
	 * @param functor
	 * @param arity
	 * @return
	 */
	private static String PredFA(String functor, int arity) {
		return "PRED_" + functor + "_" + arity;
	}

	/**
	 * @param pkg
	 * @return
	 */
	public static boolean isGlobal(String pkg) {
		return pkg != null && ("system".equals(pkg) || Prolog.BUILTIN.equals(pkg));
		// || "swi_supp".equals(pkg) ||
	}

	public static boolean isLocal(String pkg) {
		return pkg == null;
	}

	public static Operation findPredicateOp(String pkg, String functor, int arity) {
		String key = keyString(pkg, functor, arity);
		Operation op = predicateCache.get(key);
		return op;
	}

	public static void registerBuiltin(String functor, int arity, Operation cont) {
		String module = moduleFromClass(cont);
		if (arity == 0 && isEntryPoint(functor)) {
			registerFileEntryPoint(module, functor, arity, cont);
		} else {
			registerPredicate("system", functor, arity, cont);
			registerPredicate(module, functor, arity, cont, true);
		}
	}

	/**
	 * @param functor
	 * @return
	 */
	private static boolean isEntryPoint(String functor) {

		return functor != null
				&& (functor.equals("main") || functor.equals("go") || functor.equals("top") || functor.equals("$init"));
	}

	private static void registerFileEntryPoint(String module, String functor, int arity, Operation cont) {
		String key = keyString(module, functor, arity);
		if (!key.startsWith(module)) {
			key = module + "::" + key;
		}
		storePred(key, cont, false);
		if (!initsCache.containsValue(cont)) {
			System.err.println("File " + module + " entry point: " + key);
			initsCache.put(key, cont);
		}
		registerPredicate(module, functor, arity, cont);
	}

	private static String moduleFromClass(Operation op) {
		Class class1 = op.getClass();
		Class dc = class1.getDeclaringClass();
		if (dc != null) {
			class1 = dc;
		}
		String moduleName = "" + class1.getName();
		int idx = moduleName.indexOf("$$");
		if (idx > 0) {
			moduleName = moduleName.substring(0, idx);
		}
		idx = moduleName.indexOf("_");
		if (idx > 0) {
			moduleName = moduleName.substring(idx + 1);
		}

		return moduleName;
	}

	public static void registerBuiltin(String functor, String mangle, int arity, Operation cont) {
		registerBuiltin(functor, arity, cont);
		registerBuiltin(mangle, arity, cont);
	}

	public static void runInits(Prolog machine) {
		for (Iterator<Entry<String, Operation>> iterator = initsCache.entrySet().iterator(); iterator.hasNext();) {
			Entry<String, Operation> entry = iterator.next();
			String name = entry.getKey();
			Operation oper = entry.getValue();
			int steps = 0;
			System.err.println("Running: " + name + " as " + oper);
			while (oper != null && oper != Success.SUCCESS) {
				System.err.print(".");
				System.err.flush();
				steps++;
				oper = oper.exec(machine);
			}
			System.err.println("\nDone " + steps);
			System.err.flush();
		}
		System.err.flush();
	}

}

File: src/machine/SxxMachine/PredTable.cpp
using namespace std;

#include "PredTable.h"
#include "Operation.h"
#include "Prolog.h"
#include "Success.h"

namespace SxxMachine
{

unordered_map<wstring, Operation> PredTable::predicateCache = unordered_map<wstring, Operation>();
unordered_map<wstring, Operation> PredTable::hiddenCache = unordered_map<wstring, Operation>();
unordered_map<wstring, Operation> PredTable::initsCache = unordered_map<wstring, Operation>();

	void PredTable::registerPredicate(const wstring& pkg, const wstring& functor, const int& arity, Operation cont)
	{
		registerPredicate(pkg, functor, arity, cont, false);
	}

	void PredTable::registerPredicate(const wstring& pkg, const wstring& functor, const int& arity, Operation cont, const bool& forced)
	{
		if(!storePred(keyString(pkg, functor, arity), cont, forced))
		{
		}
		if(!storePred(keyString("", functor, arity), cont, false))
		{
		}
	}

	bool PredTable::storePred(const wstring& key, Operation cont, const bool& forced)
	{
		Operation op = predicateCache[key];
		if(op == nullptr)
		{
			predicateCache.emplace(key, cont);
			return true;
		} else if(op == cont)
		{
			return true;
		}
		wstring oldmodule = moduleFromClass(op);
		if(cont == nullptr)
		{

			if(forced)
			{
				System::err::println("Deleting " + key + " which is implemented at " + oldmodule);
				predicateCache.erase(key);
				return true;
			} else
			{
				System::err::println("Not deleting " + key + " which is implemented at " + oldmodule);
				return false;
			}
		}
		wstring newmodule = moduleFromClass(cont);
		{
			if(op->getClass() != op->getClass())
			{
				if(!forced)
				{
					System::err::println("Not overriding " + key + " which is implemented at " + oldmodule + " with " + newmodule);
					hiddenCache.emplace(key, cont);
					return false;
				} else
				{
				}
			}
			hiddenCache.emplace(key, op);
			predicateCache.emplace(key, cont);
			System::err::println("Overriding " + key + " which is implemented at " + oldmodule + " with " + newmodule);
			return true;
		}
	}

	void PredTable::registerWMangle(const wstring& pkg, const wstring& functor, const wstring& mangle, const int& arity, Operation cont)
	{
		registerPredicate(pkg, functor, arity, cont);
		registerPredicate(pkg, mangle, arity, cont);
	}

	wstring PredTable::keyString(const wstring& pkg, const wstring& functor, const int& arity)
	{
		if(pkg == "" || isGlobal(pkg))
		{
			return PredFA(functor, arity);
		}
		return pkg + "::" + PredFA(functor, arity);
	}

	wstring PredTable::PredFA(const wstring& functor, const int& arity)
	{
		return "PRED_" + functor + "_" + to_string(arity);
	}

	bool PredTable::isGlobal(const wstring& pkg)
	{
		return pkg != "" && ("system" == pkg || Prolog::BUILTIN == pkg);
		// || "swi_supp".equals(pkg) || 
	}

	bool PredTable::isLocal(const wstring& pkg)
	{
		return pkg == "";
	}

	Operation PredTable::findPredicateOp(const wstring& pkg, const wstring& functor, const int& arity)
	{
		wstring key = keyString(pkg, functor, arity);
		Operation op = predicateCache[key];
		return op;
	}

	void PredTable::registerBuiltin(const wstring& functor, const int& arity, Operation cont)
	{
		wstring module = moduleFromClass(cont);
		if(arity == 0 && isEntryPoint(functor))
		{
			registerFileEntryPoint(module, functor, arity, cont);
		} else
		{
			registerPredicate("system", functor, arity, cont);
			registerPredicate(module, functor, arity, cont, true);
		}
	}

	bool PredTable::isEntryPoint(const wstring& functor)
	{

		return functor != "" && (functor == "main" || functor == "go" || functor == "top" || functor == "$init");
	}

	void PredTable::registerFileEntryPoint(const wstring& module, const wstring& functor, const int& arity, Operation cont)
	{
		wstring key = keyString(module, functor, arity);
		if(!StringHelper::startsWith(key, module))
		{
			key = module + "::" + key;
		}
		storePred(key, cont, false);
		if(!initsCache.containsValue(cont))
		{
			System::err::println("File " + module + " entry point: " + key);
			initsCache.emplace(key, cont);
		}
		registerPredicate(module, functor, arity, cont);
	}

	wstring PredTable::moduleFromClass(Operation op)
	{
		type_info class1 = op->getClass();
		type_info dc = class1.getDeclaringClass();
		if(dc != nullptr)
		{
			class1 = dc;
		}
		wstring moduleName = "" + class1.getName();
		int idx = (int)moduleName.find("$$");
		if(idx > 0)
		{
			moduleName = moduleName.substr(0, idx);
		}
		idx = (int)moduleName.find("_");
		if(idx > 0)
		{
			moduleName = moduleName.substr(idx + 1);
		}

		return moduleName;
	}

	void PredTable::registerBuiltin(const wstring& functor, const wstring& mangle, const int& arity, Operation cont)
	{
		registerBuiltin(functor, arity, cont);
		registerBuiltin(mangle, arity, cont);
	}

	void PredTable::runInits(Prolog* machine)
	{
		for(unordered_map<wstring, Operation>::const_iterator iterator = initsCache.begin(); iterator != initsCache.end(); ++iterator)
		{
			unordered_map::Entry<wstring, Operation> entry = (unordered_map::Entry<wstring, Operation>) *iterator;
			wstring name = entry.getKey();
			Operation oper = entry.getValue();
			int steps = 0;
			System::err::println("Running: " + name + " as " + oper);
			while(oper != nullptr && oper != Success::SUCCESS)
			{
				System::err::print(".");
				System::err::flush();
				steps++;
				oper = oper(machine);
			}
			System::err::println("\nDone " + to_string(steps));
			System::err::flush();
		}
		System::err::flush();
	}
}
